### PR TITLE
Avoid mem safety region collapse

### PR DESCRIPTION
**Pull request by michael-emmi**
Attempting to salvage a possibly outdated fix for Issue #197 to avoid region collapse due to memory safety checking.

### CODE REVIEW

**Comment by zvonimir** [comment]
Well, some regressions are failing. I guess that would be the first thing to investigate.

File: lib/smack/MemorySafetyChecker.cpp [hunk 1]
```diff
@@ -16,66 +16,97 @@ namespace smack {
...
-  assert (memoryLeakCheckFunction != NULL && "Memory leak check function must be present.");
-  for (inst_iterator I = inst_begin(F), E = inst_end(F); I != E; ++I) {
-    if (isa<ReturnInst>(&*I)) {
-      CallInst::Create(memoryLeakCheckFunction, "", &*I);
+Function* MemorySafetyChecker::getLeakCheckFunction(Module& M) {
+  if (!leakCheckFunction.count(&M)) {
+    auto F = M.getFunction(Naming::MEMORY_LEAK_FUNCTION);
+    assert (F && "Memory leak check function must be present.");
+    leakCheckFunction[&M] = F;
+	}
```
Comment (zvonimir) [inline lib/smack/MemorySafetyChecker.cpp]: Weird/wrong indentation.

File: lib/smack/MemorySafetyChecker.cpp [hunk 2]
```diff
@@ -16,66 +16,97 @@ namespace smack {
...
+  return leakCheckFunction[&M];
+      }
```
Comment (zvonimir) [inline lib/smack/MemorySafetyChecker.cpp]: Weird/wrong indentation.

File: lib/smack/MemorySafetyChecker.cpp [hunk 3]
```diff
@@ -16,66 +16,97 @@ namespace smack {
...
+    F->addFnAttr(Attribute::AttrKind::NoUnwind);
+    safetyCheckFunction[&M] = F;
+  }
+  return safetyCheckFunction[&M];
+}
+
+void MemorySafetyChecker::insertMemoryLeakCheck(Instruction* I) {
+  auto& M = *I->getParent()->getParent()->getParent();
+  CallInst::Create(getLeakCheckFunction(M), "", I);
     }
```
Comment (zvonimir) [inline lib/smack/MemorySafetyChecker.cpp]: Weird/wrong indentation.

File: lib/smack/MemorySafetyChecker.cpp [hunk 4]
```diff
@@ -16,66 +16,97 @@ namespace smack {
...
+
+void MemorySafetyChecker::insertMemoryAccessCheck(Value* addr, Value* size, Instruction* I) {
+  auto& M = *I->getParent()->getParent()->getParent();
+  auto& C = M.getContext();
+  auto T = PointerType::getUnqual(Type::getInt8Ty(C));
+  CallInst::Create(getSafetyCheckFunction(M), {
+    CastInst::Create(Instruction::BitCast, addr, T, "", I),
+    CastInst::CreateBitOrPointerCast(size, T, "", I)
+  }, "", I);
   }
```
Comment (zvonimir) [inline lib/smack/MemorySafetyChecker.cpp]: Weird/wrong indentation.

**Comment by zvonimir** [comment]
I've just spent some time studying SVCOMP results, where this branch times out more than the develop branch, despite typically it seems splitting memory more.
One thing I discovered is that this branch adds memory checking to `__SMACK_static_init`, which for some benchmarks generated a huge number of additional memory checks as compared to the develop branch. So it would be good to avoid putting memory checks into `__SMACK_static_init`. I suspect this should be an easy fix.

**Comment by zvonimir** [comment]
Another thing I observed. The new implementation always generates the following pattern:
```
$p12 := $bitcast.ref.ref($p11);
$p13 := $bitcast.ref.ref($i2p.i64.ref(8));
call __SMACK_check_memory_safety($p12, $p13);
```
The old one generates the following:
```
$p11 := $bitcast.ref.ref($p10);
call __SMACK_check_memory_safety($p11, 8);
```
As you can see, the new implementation has an addition bitcast, instead of passing the constant directly. I am not sure how much that affects the performance.

**Comment by michael-emmi** [comment]
Good catch: this was missing a tiny piece of logic that avoids inserting the memory-safety checks in Smack-generated functions. Commit 1174fa4 fixes that, which should bring the SVCOMP results.

The additional bit cast is needed to make sure that integer-literal size arguments have the same type as pointers, so that memory-safety-checking arithmetic can be done in the arithmetic theory underlying pointers, which would typically be integers rather than bit vectors.

**Comment by zvonimir** [comment]
This looks good to me. Just waiting for SVCOMP to finish.

**Comment by zvonimir** [comment]
It is slightly better on SVCOMP results, in particular in the memory safety categories. The difference is quite small, and it times out maybe on 10 benchmarks less. Merging now.

**Pull request merged**